File: src/gequ/crawlers/download.py
import re
from pathlib import Path

import httpx
from bs4 import BeautifulSoup


class DownloadCrawler:
    def __init__(self, output_dir: str = "downloads", cookie: str = None, user_agent: str = None, timeout: int = 30):
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(exist_ok=True)
        self.cookie = cookie
        self.user_agent = user_agent or "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        self.timeout = timeout
    
    def _get_headers(self) -> dict:
        return {
            "User-Agent": self.user_agent,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
            "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
        }
    
    def _get_cookies(self) -> dict:
        if not self.cookie:
            return {}
        cookies = {}
        for cookie_str in self.cookie.split(";"):
            cookie_str = cookie_str.strip()
            if "=" in cookie_str:
                name, value = cookie_str.split("=", 1)
                cookies[name.strip()] = value.strip()
        return cookies
    
    async def get_song_page(self, song_id: int) -> BeautifulSoup:
        url = f"https://www.gequke.com/song/{song_id}"
        async with httpx.AsyncClient(timeout=self.timeout, follow_redirects=True) as client:
            resp = await client.get(
                url,
                headers=self._get_headers(),
                cookies=self._get_cookies()
            )
            resp.raise_for_status()
            return BeautifulSoup(resp.text, "lxml")
    
    def extract_song_info(self, soup: BeautifulSoup) -> dict:
        script_tags = soup.find_all("script")
        song_info = {}
        
        for script in script_tags:
            if script.string and "window.mp3_id" in script.string:
                script_text = script.string
                
                patterns = {
                    "mp3_id": r"window\.mp3_id\s*=\s*'([^']+)'",
                    "play_id": r"window\.play_id\s*=\s*'([^']+)'",
                    "mp3_title": r"window\.mp3_title\s*=\s*'([^']+)'",
                    "mp3_author": r"window\.mp3_author\s*=\s*'([^']+)'",
                    "mp3_cover": r"window\.mp3_cover\s*=\s*'([^']+)'",
                    "mp3_extra_url": r"window\.mp3_extra_url\s*=\s*'([^']+)'",
                }
                
                for key, pattern in patterns.items():
                    match = re.search(pattern, script_text)
                    if match:
                        song_info[key] = match.group(1)
                break
        
        content_lrc2 = soup.find("div", {"id": "content-lrc2"})
        if content_lrc2:
            lrc_html = content_lrc2.decode_contents()
            lrc_text = lrc_html.replace("<br/>", "\n").replace("<br />", "\n")
            song_info["lrc"] = lrc_text
        
        return song_info
    
    async def get_mp3_url(self, play_id: str, song_id: int) -> str:
        api_url = "https://www.gequke.com/api/music"
        
        headers = {
            **self._get_headers(),
            "Accept": "application/json, text/javascript, */*; q=0.01",
            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
            "X-Requested-With": "XMLHttpRequest",
            "X-Custom-Header": "SecretKey",
            "Origin": "https://www.gequke.com",
            "Referer": f"https://www.gequke.com/song/{song_id}",
        }
        
        data = f"id={play_id}&type=0"
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            resp = await client.post(
                api_url,
                data=data,
                headers=headers,
                cookies=self._get_cookies()
            )
            
            if resp.status_code != 200:
                return None
            
            result = resp.json()
            if result.get("code") == 200:
                return result["data"]["url"]
        
        return None
    
    async def download_file(self, url: str, filepath: Path):
        async with httpx.AsyncClient(timeout=self.timeout, follow_redirects=True) as client:
            resp = await client.get(
                url,
                headers={
                    "Referer": "https://www.gequke.com/",
                    "User-Agent": self.user_agent,
                },
                cookies=self._get_cookies()
            )
            resp.raise_for_status()
            
            with open(filepath, "wb") as f:
                f.write(resp.content)
    
    async def download_cover(self, cover_url: str) -> bytes:
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            resp = await client.get(cover_url)
            resp.raise_for_status()
            return resp.content
    
    def embed_mp3_metadata(self, mp3_path: Path, title: str, author: str, cover_data: bytes):
        from mutagen.mp3 import MP3
        from mutagen.id3 import ID3, TIT2, TPE1, TALB, APIC, ID3NoHeaderError
        
        try:
            audio = MP3(mp3_path, ID3=ID3)
        except ID3NoHeaderError:
            audio = MP3(mp3_path)
            audio.add_tags()
        
        audio.tags.add(TIT2(encoding=3, text=title))
        audio.tags.add(TPE1(encoding=3, text=author))
        audio.tags.add(TALB(encoding=3, text=title))
        
        if cover_data:
            audio.tags.add(APIC(
                encoding=3,
                mime='image/jpeg',
                type=3,
                desc='Cover',
                data=cover_data
            ))
        
        audio.save()
    
    async def download_song(self, song_id: int, embed_cover: bool = True) -> dict:
        result = {"success": False, "mp3_path": None, "lrc_path": None, "error": None}
        
        try:
            soup = await self.get_song_page(song_id)
        except Exception as e:
            result["error"] = f"获取歌曲页面失败: {e}"
            return result
        
        song_info = self.extract_song_info(soup)
        
        if not song_info:
            result["error"] = "无法提取歌曲信息"
            return result
        
        if not song_info.get("play_id"):
            result["error"] = "未找到 play_id，可能需要登录"
            return result
        
        title = song_info.get("mp3_title", "Unknown")
        author = song_info.get("mp3_author", "Unknown")
        
        console = None
        try:
            from rich.console import Console
            console = Console()
            console.print(f"[cyan]歌曲: {title} - {author}[/cyan]")
        except:
            pass
        
        mp3_filename = f"{title}-{author}.mp3"
        mp3_filepath = self.output_dir / mp3_filename
        
        lrc_filename = f"{title}-{author}.lrc"
        lrc_filepath = self.output_dir / lrc_filename
        
        cover_data = None
        if embed_cover and song_info.get("mp3_cover"):
            try:
                cover_data = await self.download_cover(song_info["mp3_cover"])
            except Exception as e:
                if console:
                    console.print(f"[yellow]下载封面失败: {e}[/yellow]")
        
        try:
            mp3_url = await self.get_mp3_url(song_info["play_id"], song_id)
            if not mp3_url:
                result["error"] = "API 未返回 MP3 链接"
                if song_info.get("mp3_extra_url"):
                    result["error"] += f"，备用链接: {song_info['mp3_extra_url']}"
                return result
            
            await self.download_file(mp3_url, mp3_filepath)
            
            if cover_data:
                self.embed_mp3_metadata(mp3_filepath, title, author, cover_data)
            
            result["success"] = True
            result["mp3_path"] = str(mp3_filepath)
            
        except Exception as e:
            result["error"] = f"下载 MP3 失败: {e}"
            return result
        
        if song_info.get("lrc"):
            try:
                with open(lrc_filepath, "w", encoding="utf-8") as f:
                    f.write(song_info["lrc"])
                result["lrc_path"] = str(lrc_filepath)
            except Exception as e:
                if console:
                    console.print(f"[yellow]保存歌词失败: {e}[/yellow]")
        
        return result

File: gequke-rs/src/crawlers/download.rs
use anyhow::{Result, Context};
use scraper::{Html, Selector};
use serde::{Deserialize, Serialize};
use regex::Regex;
use std::path::{Path, PathBuf};
use std::io::Write;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SongInfo {
    pub mp3_id: Option<String>,
    pub play_id: Option<String>,
    pub title: Option<String>,
    pub author: Option<String>,
    pub cover: Option<String>,
    pub lrc: Option<String>,
    pub extra_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadResult {
    pub success: bool,
    pub mp3_path: Option<String>,
    pub lrc_path: Option<String>,
    pub error: Option<String>,
    pub song_info: Option<DownloadSongInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadSongInfo {
    pub song_id: i64,
    pub title: String,
    pub artist: String,
    pub cover_url: Option<String>,
    pub play_id: Option<String>,
}

pub struct DownloadCrawler {
    output_dir: PathBuf,
    cookie: Option<String>,
    user_agent: String,
    timeout: f64,
    client: reqwest::Client,
}

impl DownloadCrawler {
    pub fn new(output_dir: &str, cookie: Option<String>, user_agent: Option<String>, timeout: Option<f64>) -> Self {
        let ua = user_agent.unwrap_or_else(|| "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36".to_string());
        let t = timeout.unwrap_or(30.0);

        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert("User-Agent", reqwest::header::HeaderValue::from_str(&ua).unwrap());
        headers.insert("Accept", reqwest::header::HeaderValue::from_static("text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8"));
        headers.insert("Accept-Language", reqwest::header::HeaderValue::from_static("zh-CN,zh;q=0.9,en;q=0.8"));

        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs_f64(t))
            .default_headers(headers)
            .redirect(reqwest::redirect::Policy::limited(10))
            .build()
            .unwrap();

        let output = PathBuf::from(output_dir);
        std::fs::create_dir_all(&output).ok();

        Self {
            output_dir: output,
            cookie,
            user_agent: ua,
            timeout: t,
            client,
        }
    }

    fn get_cookies(&self) -> Option<String> {
        self.cookie.clone()
    }

    pub async fn get_song_page(&self, song_id: i64) -> Result<Html> {
        let url = format!("https://www.gequke.com/song/{}", song_id);
        
        let mut request = self.client.get(&url);
        
        if let Some(cookie) = self.get_cookies() {
            request = request.header("Cookie", cookie);
        }

        let resp = request.send().await.context("获取歌曲页面失败")?;
        let text = resp.text().await.context("读取歌曲页面内容失败")?;
        Ok(Html::parse_document(&text))
    }

    pub fn extract_song_info(&self, doc: &Html) -> SongInfo {
        let mut info = SongInfo {
            mp3_id: None,
            play_id: None,
            title: None,
            author: None,
            cover: None,
            lrc: None,
            extra_url: None,
        };

        let script_selector = Selector::parse("script").unwrap();
        for script in doc.select(&script_selector) {
            if let Some(script_text) = script.text().next() {
                if script_text.contains("window.mp3_id") {
                    let patterns = [
                        ("mp3_id", r"window\.mp3_id\s*=\s*'([^']+)'"),
                        ("play_id", r"window\.play_id\s*=\s*'([^']+)'"),
                        ("mp3_title", r"window\.mp3_title\s*=\s*'([^']+)'"),
                        ("mp3_author", r"window\.mp3_author\s*=\s*'([^']+)'"),
                        ("mp3_cover", r"window\.mp3_cover\s*=\s*'([^']+)'"),
                        ("mp3_extra_url", r"window\.mp3_extra_url\s*=\s*'([^']+)'"),
                    ];

                    for (key, pattern) in patterns {
                        let re = Regex::new(pattern).unwrap();
                        if let Some(caps) = re.captures(script_text) {
                            let value = caps[1].to_string();
                            match key {
                                "mp3_id" => info.mp3_id = Some(value),
                                "play_id" => info.play_id = Some(value),
                                "mp3_title" => info.title = Some(value),
                                "mp3_author" => info.author = Some(value),
                                "mp3_cover" => info.cover = Some(value),
                                "mp3_extra_url" => info.extra_url = Some(value),
                                _ => {}
                            }
                        }
                    }
                    break;
                }
            }
        }

        let lrc_selector = Selector::parse("div#content-lrc2").unwrap();
        if let Some(lrc_div) = doc.select(&lrc_selector).next() {
            let lrc_html = lrc_div.inner_html();
            let lrc_text = lrc_html.replace("<br/>", "\n").replace("<br />", "\n");
            info.lrc = Some(lrc_text);
        }

        info
    }

    pub async fn get_mp3_url(&self, play_id: &str, song_id: i64) -> Result<Option<String>> {
        let api_url = "https://www.gequke.com/api/music";

        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert("Accept", reqwest::header::HeaderValue::from_static("application/json, text/javascript, */*; q=0.01"));
        headers.insert("Content-Type", reqwest::header::HeaderValue::from_static("application/x-www-form-urlencoded; charset=UTF-8"));
        headers.insert("X-Requested-With", reqwest::header::HeaderValue::from_static("XMLHttpRequest"));
        headers.insert("X-Custom-Header", reqwest::header::HeaderValue::from_static("SecretKey"));
        headers.insert("Origin", reqwest::header::HeaderValue::from_static("https://www.gequke.com"));
        headers.insert("Referer", reqwest::header::HeaderValue::from_str(&format!("https://www.gequke.com/song/{}", song_id)).unwrap());

        let body = format!("id={}&type=0", play_id);

        let mut request = self.client
            .post(api_url)
            .headers(headers)
            .body(body);

        if let Some(cookie) = self.get_cookies() {
            request = request.header("Cookie", cookie);
        }

        let resp = request.send().await.context("请求API失败")?;

        if resp.status() != 200 {
            return Ok(None);
        }

        let json: serde_json::Value = resp.json().await.context("解析API响应失败")?;
        if let Some(code) = json.get("code") {
            if code.as_i64() == Some(200) {
                if let Some(url) = json.get("data").and_then(|d| d.get("url")).and_then(|u| u.as_str()) {
                    return Ok(Some(url.to_string()));
                }
            }
        }

        Ok(None)
    }

    pub async fn download_file(&self, url: &str, filepath: &Path) -> Result<String> {
        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert("Referer", reqwest::header::HeaderValue::from_static("https://www.gequke.com/"));
        headers.insert("User-Agent", reqwest::header::HeaderValue::from_str(&self.user_agent).unwrap());

        let resp = self.client
            .get(url)
            .headers(headers)
            .send()
            .await
            .context("下载文件失败")?;

        let bytes = resp.bytes().await.context("读取文件内容失败")?;

        let mut actual_path = filepath.to_path_buf();
        let url_lower = url.to_lowercase();
        if url_lower.contains(".aac") && filepath.extension().map(|e| e != "aac").unwrap_or(false) {
            actual_path = filepath.with_extension("aac");
        } else if url_lower.contains(".flac") && filepath.extension().map(|e| e != "flac").unwrap_or(false) {
            actual_path = filepath.with_extension("flac");
        } else if url_lower.contains(".m4a") && filepath.extension().map(|e| e != "m4a").unwrap_or(false) {
            actual_path = filepath.with_extension("m4a");
        }

        let mut file = std::fs::File::create(&actual_path).context("创建文件失败")?;
        file.write_all(&bytes).context("写入文件失败")?;

        Ok(actual_path.to_string_lossy().to_string())
    }

    pub async fn download_cover(&self, cover_url: &str) -> Result<Vec<u8>> {
        let resp = self.client
            .get(cover_url)
            .send()
            .await
            .context("下载封面失败")?;

        let bytes = resp.bytes().await.context("读取封面内容失败")?;
        Ok(bytes.to_vec())
    }

    pub async fn download_song(&self, song_id: i64, embed_cover: bool) -> DownloadResult {
        let mut result = DownloadResult {
            success: false,
            mp3_path: None,
            lrc_path: None,
            error: None,
            song_info: None,
        };

        let doc = match self.get_song_page(song_id).await {
            Ok(d) => d,
            Err(e) => {
                result.error = Some(format!("获取歌曲页面失败: {}", e));
                return result;
            }
        };

        let info = self.extract_song_info(&doc);

        if info.play_id.is_none() {
            result.error = Some("未找到 play_id，可能需要登录".to_string());
            return result;
        }

        let title = info.title.clone().unwrap_or_else(|| "Unknown".to_string());
        let author = info.author.clone().unwrap_or_else(|| "Unknown".to_string());

        result.song_info = Some(DownloadSongInfo {
            song_id,
            title: title.clone(),
            artist: author.clone(),
            cover_url: info.cover.clone(),
            play_id: info.play_id.clone(),
        });

        println!("\x1b[36m歌曲: {} - {}\x1b[0m", title, author);

        let mp3_filename = format!("{}-{}.mp3", title, author);
        let mp3_filepath = self.output_dir.join(sanitize_filename(&mp3_filename));

        let lrc_filename = format!("{}-{}.lrc", title, author);
        let lrc_filepath = self.output_dir.join(sanitize_filename(&lrc_filename));

        let mut _cover_data: Option<Vec<u8>> = None;
        if embed_cover {
            if let Some(ref cover_url) = info.cover {
                match self.download_cover(cover_url).await {
                    Ok(data) => _cover_data = Some(data),
                    Err(e) => println!("\x1b[33m下载封面失败: {}\x1b[0m", e),
                }
            }
        }

        let mp3_url = match self.get_mp3_url(info.play_id.as_ref().unwrap(), song_id).await {
            Ok(Some(url)) => url,
            Ok(None) => {
                let mut error_msg = "API 未返回音频链接".to_string();
                if let Some(extra) = info.extra_url {
                    error_msg.push_str(&format!("，备用链接: {}", extra));
                }
                result.error = Some(error_msg);
                return result;
            }
            Err(e) => {
                result.error = Some(format!("获取MP3链接失败: {}", e));
                return result;
            }
        };

        match self.download_file(&mp3_url, &mp3_filepath).await {
            Ok(actual_path) => {
                result.mp3_path = Some(actual_path);
                result.success = true;
            }
            Err(e) => {
                result.error = Some(format!("下载 MP3 失败: {}", e));
                return result;
            }
        }

        if let Some(lrc) = info.lrc {
            if !lrc.is_empty() {
                match std::fs::write(&lrc_filepath, lrc) {
                    Ok(_) => result.lrc_path = Some(lrc_filepath.to_string_lossy().to_string()),
                    Err(e) => println!("\x1b[33m保存歌词失败: {}\x1b[0m", e),
                }
            }
        }

        result
    }
}

fn sanitize_filename(filename: &str) -> String {
    let invalid_chars = ['<', '>', ':', '"', '/', '\\', '|', '?', '*'];
    let mut result = String::new();
    for c in filename.chars() {
        if !invalid_chars.contains(&c) {
            result.push(c);
        }
    }
    result
}